### ISSUE TITLE
Zoom and rotatión features

### ISSUE THREAD

**Issue by Velkamhell97** [deakjahn/crop_image#11]
This is a awesome package, is the only that allows to custom the UI as you want, please if you could add the zoom and rotate methods to the controller it would be perfect, great work, thanks for your dedication.

**Comment by deakjahn** [comment]
Well, I'll think about it. :-)

**Comment by monsieurtanuki** [comment]
@deakjahn For the record I've implemented a 0-90-180-270 rotation tool based on your code, in https://github.com/openfoodfacts/smooth-app/pull/2872.
If you're interested feel free to find inspiration there or to tell me if you'd like me to code a PR.

**Comment by deakjahn** [comment]
Yes, sure, why not. I couldn't yet take a closer look, is this something that would be easy to incorporate, with retaining existng functionality, without breaking changes?

**Comment by monsieurtanuki** [comment]
Quickly said, the differences I see:
* `crop_grid.dart`: as far as I remember, identical
* `rotated_crop_controller.dart`: additional `Rotation` parameter; the `croppedBitmap` returns an image (from `image/image.dart`)
* `rotated_crop_image.dart`: now I use a `CustomPainter` to display the rotated photo (I don't think you used one), and for that I think I changed the image constructor parameter (now it's `final ui.Image image`), which is probably a breaking change

**Comment by deakjahn** [comment]
For the last one I would then suggest a new, named constructor. We can avoid the breaking change that way, besides if somebody doesn't need the rotation at all, they would see no difference at all.

**Comment by monsieurtanuki** [comment]
Indeed. What's the next step now?

**Comment by deakjahn** [comment]
Well, the easiest for me would be if you could submit a PR that does everything. ;-)

**Comment by monsieurtanuki** [comment]
Makes sense, unfortunately ;)
I'll send a PR within a week.
I'm not sure how integrated the current crop and "my" rotation crop can be. We'll see.

**Comment by monsieurtanuki** [comment]
@deakjahn Sorry for the delay; I should be able to PR this week.
Beyond the rotation, there's an additional feature: giving enough space to corners so that their related touch area is as big as if in the middle of the image.
Not related: what's the name of the kitten?

| tool | capture |
| -- | -- |
| ![Screenshot_2023-02-14-19-32-18](https://user-images.githubusercontent.com/11576431/218827014-1193d50c-8f54-48be-857a-4b79c1ce58da.png) | ![Screenshot_2023-02-14-19-34-22](https://user-images.githubusercontent.com/11576431/218827000-9f14a135-662c-4fd3-b639-8a90a9013e2a.png) |

**Comment by deakjahn** [comment]
OK, waiting for it.

There was no name. The story is a bit longer than with a regular house cat. I live in the city but in a district with a lot of green areas. Back then, it was about ten years ago, we regularly had a few stray cats living in these gardens, roaming freely (I actually don't know if they were stray or feral in the strictest sense of the word but certainly not wild in any way, just free). I decided to feed one regularly, every day, and we became good friends. This was the mother. And she disappeared one day for a month or so, giving birth to kittens (probably in the cellar of the neighboring house but I'm not sure). She was so friendly, well tempered, calm and trusting that when she first appeared with five kittens in our garden, she simply sat aside and let me, and later other people, including children handle the kittens and play with them without any fear.

I made hundreds of pictures and short videos back then in those weeks, with the intention to collect enough good ones for a printed year calendar on the wall and I actually did, this one being one of the selected 12+1 best pictures at the end. The five kittens then went to different people, we struggled to find new homes for them. This one looked the weakest of the litter, but actually, I received a message back about the strongest one that he (I seem to recall that was a tomcat but I'm not sure any more) didn't live very long later. I have no idea of the fate of the rest.

Anyway, we brought the mother to the vet the next time around, to avoid more kittens, and it was a good thing to do because, as the vet said, there were so many kittens the second time that she would have been unable to give birth and survive the kittening. A friend accepted her during the required few rest days after the operation, and eventually adopted her. She, being stray or feral or whatever, very soon adapted and become a real house cat, literally in a few days' time, probably seeing how much better of a life she had there :-), but as I said, she was of a very calm and well tempered nature to start with. And the surroundings were even more favorable, a detached family house with even more green and free space to roam in the neighborhood. She lived quite a few years there, and although we never were sure of her precise age (the wet said it was much more difficult to tell for cats than for dogs), she probably lived a near normal lifespan.

But I still have the large selection of pictures and use them whenever I have a need for sample images. :-)

**Comment by monsieurtanuki** [comment]
Thank you @deakjahn for that touching story! Now the kitten will live forever, in github.

If I go back to the curent issue, I'm confronting different problems:
* adding aspect ratio to rotation + minimum crop size is a bit complex - it's just a question of me being lazy fixing it (so it's just a matter of time)
* in my own project I need to be immediately aware of the crop changes, like listening to `_CropControllerValue` - I think it would be interesting in general (not only for rotation) but would need some refactoring and probably (minor) breaking changes. That would mean a version 2.0.0 - do you agree with that?

**Comment by monsieurtanuki** [comment]
@deakjahn I think I found a lighter solution based on `_controller.addListener(() {})`, so this will probably not be a breaking change after all.

**Comment by deakjahn** [comment]
@monsieurtanuki If you found a solution, sure, or I would also think about a stream that sends some kind of events and then anybody can subscribe to it if interested.

**Comment by PavieOlivier** [comment]
@monsieurtanuki  Any updates by any chance?
Did you push your code on a separate branch?

**Comment by monsieurtanuki** [comment]
Hi @PavieOlivier!
My latest PR was merged here: the 90 degree rotations are available.
The only thing problematic - according to @deakjahn and I can understand it - is that the crop area gets smaller and smaller when you're rotating with a fixed spect ratio.

**Comment by monsieurtanuki** [comment]
To me the quick fix  (solution 1) is "if you rotate while using an aspect ratio, after rotation you get the biggest crop area with this aspect ratio". I could code that within a week I guess - but I'm a bit busy currently.

If the fix is (solution 2) "I want the same number of pixels (image pixels? screen pixels?) in the crop areas of all rotations", I'm not sure I would be able to code that quickly, and I would prefer not to.

As an end user I can understand that the current solution (shrinking crop areas) is not good.
Solution 1 is consistent ("always max size") and as I say, "good enough is good enough".

@deakjahn @PavieOlivier Do you agree with solution 1?

**Comment by deakjahn** [comment]
Yes, sure. It's unlikely to be a legitimate user operation to turn it round and round, just that we don't offer up an example that comes back with bug reports about the crop disappearing... :-)

**Comment by PavieOlivier** [comment]
Yes, 
I think solution 1 will be better

**Comment by monsieurtanuki** [comment]
@deakjahn @PavieOlivier cf #19

**Comment by monsieurtanuki** [comment]
Hi @deakjahn!

Just wondering about the zoom.

Perhaps adding methods `zoomIn` and `zoomOut` to the controller would be good enough (I mean, no fancy animations, no "candy crush" features):
* `zoomIn`: image zoomed and centered on the current crop rect. Perhaps with image scroll.
* `zoomOut`: zoom centered on the full 0,0,1,1 rect. Business as usual.
